### PR TITLE
feat: add ability to toggle auctions on and off

### PR DESCRIPTION
**Pull request by coreyar**
## Description

Adds toggle to turn auctions on and off.

### CODE REVIEW

**Review by chechu — COMMENTED** [review]
We could follow an approach similar to what we have done in the Vault contracts [here](https://github.com/VenusProtocol/venus-protocol/pull/236), making them "pausable". It uses a modifier and define the guard variable as "pause" (opposite to "enabled")

**Comment by chechu** [comment]
> We could follow an approach similar to what we have done in the Vault contracts [here](https://github.com/VenusProtocol/venus-protocol/pull/236), making them "pausable". It uses a modifier and defines the guard variable as "pause" (opposite to "enabled")

And we could extend this "pause" barrier to other functions: placeBid, closeAuction, restartAuction

**Comment by coreyar** [comment]
> > We could follow an approach similar to what we have done in the Vault contracts [here](https://github.com/VenusProtocol/venus-protocol/pull/236), making them "pausable". It uses a modifier and defines the guard variable as "pause" (opposite to "enabled")
> 
> And we could extend this "pause" barrier to other functions: placeBid, closeAuction, restartAuction

I thought about this and decided that it isn't necessary to "pause" those functions.

If an auction is currently in progress we should let it finish which means bid and close still need to work if auctions are paused. Once auctions are paused `placeBid` and `closeAuction` will no longer work because there will not be any auctions. `restartAuction` won't work because it calls `startAuction`.

If we want to stop in progress auctions we need to add extra logic to send the locked funds back and end the auction

**Comment by coreyar** [comment]
> > > We could follow an approach similar to what we have done in the Vault contracts [here](https://github.com/VenusProtocol/venus-protocol/pull/236), making them "pausable". It uses a modifier and defines the guard variable as "pause" (opposite to "enabled")
> > 
> > 
> > And we could extend this "pause" barrier to other functions: placeBid, closeAuction, restartAuction
> 
> I thought about this and decided that it isn't necessary to "pause" those functions.
> 
> If an auction is currently in progress we should let it finish which means bid and close still need to work if auctions are paused. Once auctions are paused `placeBid` and `closeAuction` will no longer work because there will not be any auctions. `restartAuction` won't work because it calls `startAuction`.
> 
> If we want to pause in progress auctions, I think we need to add extra logic to send the locked funds back and end the auction because it isn't right for us to lock those funds until the auctions are unpaused again.

Auctions will be paused and unpaused via VIP so participants will have time to react before pausing.

File: contracts/Shortfall/Shortfall.sol [hunk 1]
```diff
@@ -354,6 +371,25 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlled, ReentrancyGuard
...
+     * @custom:access Restricted by ACM
+     */
+    function pauseAuctions() external {
+        _checkAccessAllowed("pauseAuctions()");
+        require(auctionsPaused == false, "Auctions are already paused");
+        auctionsPaused = true;
+        emit AuctionsPaused(msg.sender);
+    }
+
+    function resumeAuctions() external {
```
Comment (GitGuru7) [inline contracts/Shortfall/Shortfall.sol:386]: Likely above we can add netspec comments in this external function

**Review by defcon022 — COMMENTED** [review]
Tests are failing.

File: contracts/Shortfall/Shortfall.sol [hunk 2]
```diff
@@ -354,6 +371,25 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlled, ReentrancyGuard
...
     }
 
+    /**
+     * @notice Toggle whether auctions are enabled or disabled
+     * @custom:event Emits AuctionEnabledToggled on success
+     * @custom:access Restricted by ACM
+     */
+    function pauseAuctions() external {
+        _checkAccessAllowed("pauseAuctions()");
+        require(auctionsPaused == false, "Auctions are already paused");
```
Comment (chechu) [inline contracts/Shortfall/Shortfall.sol:381]:
```suggestion
        require(!auctionsPaused, "Auctions are already paused");
```

File: contracts/Shortfall/Shortfall.sol [hunk 3]
```diff
@@ -354,6 +371,25 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlled, ReentrancyGuard
...
+        auctionsPaused = true;
+        emit AuctionsPaused(msg.sender);
+    }
+
+    function resumeAuctions() external {
+        _checkAccessAllowed("resumeAuctions()");
+        require(auctionsPaused == true, "Auctions are not paused");
```
Comment (chechu) [inline contracts/Shortfall/Shortfall.sol:388]:
```suggestion
        require(auctionsPaused, "Auctions are not paused");
```

File: contracts/Shortfall/Shortfall.sol [hunk 4]
```diff
@@ -261,15 +275,18 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlled, ReentrancyGuard
      * @notice Start a auction when there is not currently one active
      * @param comptroller Comptroller address of the pool
      * @custom:event Emits AuctionStarted event on success
+     * @custom:event Errors if auctions are paused
      */
     function startAuction(address comptroller) external {
+        require(auctionsPaused == false, "Auctions are paused");
```
Comment (chechu) [inline contracts/Shortfall/Shortfall.sol:281]:
```suggestion
        require(!auctionsPaused, "Auctions are paused");
```

File: contracts/Shortfall/Shortfall.sol [hunk 5]
```diff
@@ -364,6 +378,32 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlledV8, ReentrancyGua
         emit PoolRegistryUpdated(oldPoolRegistry, poolRegistry_);
     }
 
+    /**
+     * @notice Pause auctions. This disables starting new auctions but lets the current auction finish.s
```
Comment (chechu) [inline contracts/Shortfall/Shortfall.sol:382]:
```suggestion
     * @notice Pause auctions. This disables starting new auctions but lets the current auction finishes
```

File: tests/hardhat/Shortfall.ts [hunk 6]
```diff
@@ -614,6 +616,29 @@ describe("Shortfall: Tests", async function () {
...
+        .to.emit(shortfall, "AuctionsPaused")
+        .withArgs(owner.address);
+      // vDAI.badDebt.returns(parseUnits("10000", 18));
+      // await vDAI.setVariable("badDebt", parseUnits("10000", 18));
+      // vWBTC.badDebt.returns(parseUnits("2", 8));
+      // await vWBTC.setVariable("badDebt", parseUnits("2", 8));
+
+      // await expect(shortfall.startAuction(poolAddress)).to.be.revertedWith("Auctions are paused");
+      // await expect(shortfall.closeAuction(poolAddress)).to.be.revertedWith("Auctions are paused");
+      // await expect(shortfall.placeBid(poolAddress, 1)).to.be.revertedWith("Auctions are paused");
```
Comment (chechu) [inline tests/hardhat/Shortfall.ts:641]: ?

**Review by chechu — APPROVED** [review]

**Pull request merged**